tests/Benchmarks/System.Text.JsonLab/JsonWriterPerf.cs

```diff
@@ -100,7 +102,7 @@ private TextWriter GetWriter()
             return writer;
         }
 
-        private static void WriterSystemTextJsonBasicUtf8(bool formatted, ArrayFormatter output, int[] data)
+        private static void WriterSystemTextJsonBasicUtf8(bool formatted, BufferWriter<IBufferWriter<byte>> output, int[] data)
```

Comment:
Again add a convenience helper
```csharp
public ref partial struct BufferWriter
{
    public static BufferWriter<T> Create<T>(TBufferWriter bufferWriter) where TBufferWriter : IBufferWriter<byte>
    {
        return new BufferWriter<TBufferWriter>(bufferWriter);
    }
}
```
Though maybe they should be called `BufferedWriter` and `BufferedWriter<TBufferWriter>`?

```csharp
private static void WriterSystemTextJsonBasicUtf8<TBufferWriter>(bool formatted, TBufferWriter output, int[] data)
    where TBufferWriter : IBufferWriter<byte>
{
    WriterSystemTextJsonBasicUtf8(formatted, BufferWriter.Create(output), data);
}

private static void WriterSystemTextJsonBasicUtf8(bool formatted, BufferWriter<TBufferWriter> output, int[] data)
    where TBufferWriter IBufferWriter<byte>
{
    var  json = JsonWriter.Create(output, isUtf8: true, formatted);
    // ...
```

Reply:
This is good feedback specific to BufferWriter.
https://github.com/dotnet/corefxlab/issues/2177

cc @KrzysztofCwalina, @JeremyKuhne

Reply:
Added comment for issue

Reply:
Change to overload is still relevant for this PR though

Reply:
> Again add a convenience helper


We already have such a convenience method:
https://github.com/dotnet/corefxlab/blob/a02d39de49df0257b9b3a92bc7504ede33847f1e/src/System.Buffers.ReaderWriter/System/Buffers/Writer/BufferWriter.cs#L24-L26

Reply:
Concrete type it, but not its parameter
```csharp
public static class JsonWriter
{
    public static JsonWriter<TBufferWriter> Create<TBufferWriter>(
        TBufferWriter bufferWriter,
        bool isUtf8,
        bool prettyPrint = false)
    where TBufferWriter : IBufferWriter<byte>
    {
        return new JsonWriter<TBufferWriter>(BufferWriter.Create(bufferWriter), isUtf8, prettyPrint);
    }

    // Stream overload
    public static JsonWriter<StreamWriter> Create(
        Stream stream,
        bool isUtf8,
        bool prettyPrint = false)
    {
        return new JsonWriter<StreamWriter>(BufferWriter.Create(new StreamWriter(stream)), isUtf8, prettyPrint);
    }
}

public ref struct JsonWriter<TBufferWriter>
    where TBufferWriter : IBufferWriter<byte>
{
    private BufferWriter<TBufferWriter> _bufferWriter;

    public JsonWriter(BufferWriter<TBufferWriter> bufferWriter, bool isUtf8, bool prettyPrint = false)
    {
        _bufferWriter = bufferWriter;
        // ...
    }
}
```

Reply:
> Concrete type it, but not its parameter?

I don't follow. We still have the issue with JsonWriter.Create since the first argument a generic type:
```C#
var formatter = new ArrayFormatter(1024, SymbolTable.InvariantUtf8);
var bufferWriter = BufferWriter.Create(formatter);
var json = JsonWriter.Create(bufferWriter, isUtf8: true, prettyPrint: false); // Build error
```

Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0306	The type `BufferWriter<ArrayFormatter>` may not be used as a type argument	System.Text.JsonLab.Tests	E:\GitHub\Fork\corefxlab\tests\System.Text.JsonLab.Tests\JsonWriterTests.cs	23	Active

Reply:
Ah, gotcha. Just need to add concrete Create overload
```csharp
public static class JsonWriter
{
    public static JsonWriter<TBufferWriter> Create<TBufferWriter>(
        TBufferWriter bufferWriter,
        bool isUtf8,
        bool prettyPrint = false)
    where TBufferWriter : IBufferWriter<byte>
    {
        return new JsonWriter<TBufferWriter>(BufferWriter.Create(bufferWriter), isUtf8, prettyPrint);
    }

    // extra overload
    public static JsonWriter<TBufferWriter> Create<TBufferWriter>(
        BufferWriter<TBufferWriter> bufferWriter,
        bool isUtf8,
        bool prettyPrint = false)
    where TBufferWriter : IBufferWriter<byte>
    {
        return new JsonWriter<TBufferWriter>(bufferWriter, isUtf8, prettyPrint);
    }
}
```